### PR TITLE
Add a type of format strings to Ltac2.

### LINKED ISSUE #10108: Message.printf
The verbosity of `Message.of_*` and ` Message.concat ` is the third most annoying source of verbosity.  I'd like to see `Message.printf` with the various format specifiers.  (I guess there's, what?, `%a` for custom, `%s ` for string, `%d` for int, `%x`(?) for int as hex, and maybe also `%c` for constr, `%i ` for indent, `%m` for message, `%b` for Boolean (or is this used for int as binary?)?)  It might be nice to be able to declare / add format specifiers in user-space, but even one that just handled string, constr, and custom would be very welcome.

### PR DESCRIPTION
**Pull request by ppedrot**
It provides an abstract type of well-typed format strings, a scope to parse them and a minimal printf-like API.

Fixes / closes #10108

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by herbelin** [comment]
Cool.

**Comment by herbelin** [comment]
This reminds me that there are Coq versions of `printf` around. This should probably be provided in the/a standard library.

**Comment by ppedrot** [comment]
So this should be ready now. I have documented the macro and the scope syntax in the Printf module, but I haven't added anything to the Ltac2 documentation proper. I am not sure it is worth the duplication, and where to put it to start with. Maybe a reference to the format scope in the list of built-in scopes?

File: doc/changelog/05-tactic-language/13236-ltac2-printf.rst
Comment (JasonGross) [inline doc/changelog/05-tactic-language/13236-ltac2-printf.rst:3]:
```suggestion
  Added a ``printf`` macro to Ltac2. It can be made accessible by
  importing the ``Ltac2.Printf`` module. See the documentation
```

**Review by JasonGross — APPROVED** [review]
LGTM.  I am excited for this change to land!

I do think at the very least a table of the valid specifiers and their meanings should make its way into the reference manual chapter on Ltac2 eventually, but I'm fine with not blocking this PR for it.  (cc @coq/doc-maintainers ?)

Non-blocking suggestion: Make the test-suite file an output test file.

Further thoughts: Is there any hope of making the format specifiers extendable via plugin or, better, vernacular?  It would be nice to be able to register custom data structures as having dedicated format specifiers in a way that makes them as first-class as the built-in ones.

**Comment by jfehrle** [comment]
> I do think at the very least a table of the valid specifiers and their meanings should make its way into the reference manual chapter on Ltac2 eventually, but I'm fine with not blocking this PR for it.

Yeah, we shouldn't block this PR for lack of documentation.  On the other hand, undocumented features in effect don't exist for users--how are they supposed to learn about the added functionality?  How about at least adding a sentence pointing to the partial description of the feature in `Printf.v` (maybe under "Standard IO" and tweaking (or removing) statements in the doc such as "Printing functions are limited and awkward to use. Only a few data types are printable."?

Adding a table in the doc would be great.  The overall organization of the Ltac2 chapter is a bit of a jumble and will need considerable reorganization, so I wouldn't worry too much about where to put it.  Maybe below "Standard IO"?  Having the info in the doc even if it's not so easy to find would be a good step forward.

**Comment by Zimmi48** [comment]
>Yeah, we shouldn't block this PR for lack of documentation.

No documentation in the reference manual doesn't mean no documentation at all. This PR *is* documented (https://coq.gitlab.io/-/coq/-/jobs/977727260/artifacts/_install_ci/share/doc/coq/html/stdlib/Ltac2.Printf.html). If a PR adds undocumented features, then yes, by all means, it should be blocked from being merged and the exceptions to this rule should be very few.

**Comment by kyoDralliam** [comment]
There seems to be a small problem in the way that documentation is displayed on the link @Zimmi48 shared:
```
Printing specifications are of the form
': outputs 
```

**Comment by ppedrot** [comment]
It seems coqdoc doesn't like `%`, because it's the LaTeX escape. Anybody willing to write a coqdoc frontend for the Sphinx syntax?

**Comment by ppedrot** [comment]
I read both the coqdoc doc and the code and didn't find a way to escape `%` without fiddling with line arguments, so I simply wrapped the offending code in a verbatim block. Let's see how it fares.

**Comment by ppedrot** [comment]
It seems to have worked: https://coq.gitlab.io/-/coq/-/jobs/980004160/artifacts/_install_ci/share/doc/coq/html/stdlib/Ltac2.Printf.html.

**Comment by jfehrle** [comment]
> No documentation in the reference manual doesn't mean no documentation at all. This PR is documented (https://coq.gitlab.io/-/coq/-/jobs/977727260/artifacts/_install_ci/share/doc/coq/html/stdlib/Ltac2.Printf.html). If a PR adds undocumented features, then yes, by all means, it should be blocked from being merged and the exceptions to this rule should be very few.

Only being in the standard library is almost the same as being undocumented IHMO.  It's in there somewhere if you can manage to find it.  The overall organization of the standard library pages are so weak.  PRs making changes to aren't flagged for review by `@coq/doc-maintainers`--that's rather weak.  The standard library document is not mentioned in the reference manual, hyperlinking between them is not possible AFAIK.

**Comment by SkySkimmer** [comment]
@coq/ltac2-maintainers this needs an assignee

**Comment by MSoegtropIMC** [comment]
I can take it from here - give me a few days for the review and merge.

**Comment by MSoegtropIMC** [comment]
> Only being in the standard library is almost the same as being undocumented IHMO. It's in there somewhere if you can manage to find it. The overall organization of the standard library pages are so weak. PRs making changes to it aren't flagged for review by `@coq/doc-maintainers`--that's rather weak. The standard library document is not mentioned in the reference manual, hyperlinking between them is not possible AFAIK.

I don't agree with this view. Although it does add a few Ltac2 intrinsics, it mostly adds to the Ltac2 standard library. And the Ltac2 standard library is most naturally documented in the standard library documentation.

**Review by MSoegtropIMC — APPROVED** [review]
Less to review than a I thought - very nice and most useful! I am looking forward to it.

I will merge this tomorrow unless there is a veto on the doc question. As I wrote I see the library docs as the appropriate place to document this - sure it needs some support in the language to parse the format strings, but this is more an internal implementation detail users are not supposed to know about. In OCaml one wouldn't consider printf to be part of the language either and if one wants to understand who it works, one has to reverse engineer it as well.

**Comment by jfehrle** [comment]
> I will merge this tomorrow unless there is a veto on the doc question.

Not from me.  My comment was a general observation, not a blocker for this PR.

**Pull request merged**
